openvdb/points/PointSample.h

```diff
@@ -0,0 +1,379 @@
+         bool Transform = false>
+class PointSampleOp
+{
+public:
+    using LeafManagerT = tree::LeafManager<PointDataTreeT>;
+    using PointDataLeafNodeT = typename PointDataTreeT::LeafNodeType;
+    using PointIteratorT = IndexIter<points::ValueVoxelCIter, PointFilterT>;
+
+    using PositionAttributeHandleT = AttributeHandle<PositionT, CodecT>;
+    using TargetAttributeHandleT = AttributeWriteHandle<TargetValueT>;
+
+    PointSampleOp(const AccessorT& sourceAccessor,
+                  const size_t positionAttributeIndex,
+                  const size_t targetAttributeIndex,
+                  const PointFilterT& filter = PointFilterT(),
+                  const math::Transform* const pointDataGridTransform = nullptr,
+                  const math::Transform* const sourceGridTransform = nullptr)
+    : mSourceAccessor(sourceAccessor)
+    , mPositionAttributeIndex(positionAttributeIndex)
+    , mTargetAttributeIndex(targetAttributeIndex)
+    , mFilter(filter)
+    , mPointDataGridTransform(pointDataGridTransform)
+    , mSourceGridTransform(sourceGridTransform)
+    {
+        if (Transform) {
+            assert(mSourceGridTransform);
+            assert(mPointDataGridTransform);
+        }
+    }
+
+    void operator()(const typename LeafManagerT::LeafRange& range) const
+    {
+        PointFilterT filterCopy(mFilter);
+
+        for (auto leaf = range.begin(); leaf; ++leaf) {
+            // get the relevant point data attribute arrays for the voxel in question
+            typename PositionAttributeHandleT::Ptr positionHandle =
+                PositionAttributeHandleT::create(
+                    leaf->constAttributeArray(mPositionAttributeIndex));
+
+            typename TargetAttributeHandleT::Ptr targetHandle =
+                TargetAttributeHandleT::create(
+                    leaf->attributeArray(mTargetAttributeIndex));
+
+            filterCopy.reset(*leaf);
+
+            for (auto voxel = leaf->cbeginValueOn(); voxel; ++voxel) {
+                const Coord& coord = voxel.getCoord();
+
+                PointIteratorT iter(leaf->beginValueVoxel(coord), filterCopy);
+                if (!iter) continue;
+
+                const Vec3d coordVec3d = coord.asVec3d();
+
+                for (; iter; ++iter) {
+                    const Index index = *iter;
+
+                    // get the position in index space. Do this in double precision to make sure
+                    // we doesn't move points which are close to the -0.5/0.5 boundary into other
+                    // voxels when doing a staggered offset by 0.5
+
+                    Vec3d position = Vec3d(positionHandle->get(index)) + coordVec3d;
+
+                    // transform it to source index space if transforms differ
+
+                    if (Transform) {
+                        Vec3d worldPosition = mPointDataGridTransform->indexToWorld(position);
+                        position = mSourceGridTransform->worldToIndex(worldPosition);
+                    }
```

Comment:
I think I'd suggest wrapping up all the position access logic into a new struct so this code can be cleaned up a bit.